Xamarin.Forms.Core/Element.cs

```diff
@@ -208,9 +208,37 @@ public Element Parent
 				OnParentSet();
 
 				OnPropertyChanged();
+
+				if (this.IsTemplateRoot)
+					this.TemplatedParent = value;
 			}
 		}
 
+		internal event EventHandler TemplatedParentChanged;
+		
+		BindableObject _templatedParent;
+		public BindableObject TemplatedParent
+		{
+			get => _templatedParent ?? (_templatedParent = this.Parent?.TemplatedParent);
```

Comment:
this will not behave well with reparenting. I don't think this value should be cached when computed
```suggestion
			get => _templatedParent ?? (Parent?.TemplatedParent);
```

Reply:
I have it clearing the cached template parent for all the `Descendants()` when the actual parent changes, so that should cover reparenting, no? There's one or two older unit tests that test for this too and I made sure they passed.

The reason I wanted to cache it was if folks (like me!) want to make more use of templated controls over native controls it would be a lot more efficient to apply bindings on 100s or 1000s of controls if we didn't have to crawl the tree every time we want to get the templated parent.

Reply:
ok